src/quantuminspire/qiskit/circuit_parser.py

```diff
@@ -191,60 +431,135 @@ def _u3(self, circuit: Dict[str, Any]) -> str:
+        pass
 
-    def _ry(self, circuit: Dict[str, Any]) -> str:
-        """ Translates the Ry element.
+    @staticmethod
+    def _measure(stream: StringIO, instruction: QobjItem) -> None:
+        """ Translates the measure element. No cQASM is added for this gate.
 
         Args:
-            circuit: The Qiskit circuit with Ry element.
+            stream: The string-io stream to where the resulting cQASM is written.
+            instruction: The Qiskit instruction to translate to cQASM.
 
-        Returns:
-            cQASM code string.
         """
-        return self.__r(circuit, 'y')
+        pass
+
+    def _parse_bin_ctrl_gate(self, stream: StringIO, instruction: QobjItem) -> None:
+        """ Parses a binary controlled gate. A binary controlled gate name is preceded by 'c-'.
+            The gate is executed when a specific measurement is true. Multiple measurement outcomes are used
+            to control the quantum operation. This measurement is a combination of classical bits being 1 and others
+            being 0. Because cQASM only supports measurement outcomes of 1, any other bits in the
+            masked bit pattern first have to be inverted with the not-operator. The same inversion also has to
+            take place after the binary controlled quantum operation.
+            The resulting stream will be expanded with something like:
+            not b[the 0-bits changed to 1]
+            c-gate [classical bits], other arguments
+            not b[the 0-bits reset to 0 again]
+            When the c-gate results in an empty string (e.g. binary controlled u(0, 0, 0) or barrier gate),
+            nothing is added to the stream.
 
-    def _rz(self, circuit: Dict[str, Any]) -> str:
-        """ Translates the Rz element.
+        Args:
+            stream: The string-io stream to where the resulting cQASM is written.
+            instruction: The Qiskit instruction to translate to cQASM.
+
+        """
+        conditional_type = instruction.conditional.type
+        if conditional_type != 'equals':
+            raise ApiError('Conditional statement with type {} not supported'.format(conditional_type))
+        mask = int(instruction.conditional.mask, 16)
+        if mask == 0:
+            raise ApiError('Conditional statement {} without a mask'.format(instruction.name.lower()))
+        nr_of_classical_bits = int(np.log2(mask + 1))
+        val = int(instruction.conditional.val, 16)
+        masked_val = mask & val
+
+        # form the negation to the 0-values of the measurement registers, when mask == val no bits are negated
+        negate_zeroes_line = ''
+        if masked_val != mask:
+            negate_zeroes_line = 'not b[' + ','.join(
+                str(i) for i in range(nr_of_classical_bits) if not (masked_val & (1 << i))) + ']\n'
+
+        # form multi bits control - qasm-single-gate-multiple-qubits
+        binary_control = 'b[0:{}], '.format(nr_of_classical_bits - 1)
```

Comment:
does binary control always start at index '0'? is this not allowed `C-H b[2:3], q[1]`? And, if `nr_of_classical_bits` is 1, then this will become `b[0:0]` which is not very pretty and can not be executed from the web interface.

Reply:
binary control in QX is always on all classical bits, therefore starting with 0.
In qiskit you have something like gate.c_if(c, 6) where the integer value actually means some bits must be 1 and others must be 0. Because qx can only check bits being true, we have to inverse the bits that has to be zero and check them all for being 1, so starting from 0:n for n+1 classical bits.
I know b[0:0] generates a syntax error from the web interface but I think that is an omission in the web interface. I first had the check for a one bit program, but deleted it later, because it is legal and one conditional path to maintain less. The spec does say nothing about i < j for the construction b[i:j]. There are a few other constructions that the web interface cannot handle (for example C-Rx b[0], q[0], 3.14, also all controlled multi qubit gates are not parsed correctly by the web interface) but QX is compatible with the spec.

Reply:
> binary control in QX is always on all classical bits, therefore starting with 0.

This is incorrect. Constructions like `c-X b[3], b[5], q[2]` are perfectly legal in cqasm (and thus in qx).

Also, if you negate classical bits for a particular c-? instruction, you should negate them back to their original value in case those same bits are used again later.

Reply:
I meant sent to QX as in constructed by the SDK. Changed it.

Reply:
In qiskit the gate:
`X(q[2]).c_if(c, 40)`
cannot be simply translated in:
`c-X b[3], b[5], q[2]`
because int(40) is bin(101000). The zero bits must be checked too. in cQASM classical bits can only be check being positive. That's why we have to negate the zero-bits before and after the conditional gate and checking all the classical bits as:
`not b[zero bits]
c-X b[3], b[5], q[2]
not b[zero bits]`